\documentclass{article}

\usepackage[margin=1in]{geometry}
\usepackage{graphicx}

\title{Specification for a Crux graphical user interface}

\author{
Charles Grant\\
Department of Genome Sciences\\
Department of Computer Science and Engineering\\
University of Washington
\and
William Stafford Noble\\
Department of Genome Sciences\\
Department of Computer Science and Engineering\\
University of Washington}

\begin{document}

\maketitle

\section{Requirements}
\subsection{Introduction}

Crux is a software toolkit for analyzing shotgun proteomics data.
Currently, users apply these tools via a command line interface defined
with respect to a single binary executable.  The toolkit includes
tools for searching a sequence database, assigning statistical
confidence estimates to the resulting peptide-spectrum matches, and
inferring protein identifications from a collection of spectra.

We would like to provide a graphical user interface (GUI) that is
intuitive, flexible and portable.  Logistically, we have the following
design goals:
\begin{itemize}
\item Crux should run under Linux, MacOS and Windows.
\item Crux should be easy to install.
\item Any code that we use must be compatible with the Crux license.
\end{itemize}
More philosophical goals include the following:
\begin{itemize}
\item The Crux GUI should provide an easy way for a novice user to
  carry out common tasks, as well as flexibility for advanced users to
  perform more esoteric tasks.
\item For a multi-step analysis, a mechanism should be provided to
  allow the user to easily re-use steps from previous analyses.
\item Crux will not provide the functionality of a full laboratory
  information management system.
\end{itemize}

\subsection{A simple analysis}

\begin{figure}
\centering
\includegraphics[width=6in]{schematic.eps}
\caption{{\bf The Crux toolkit.}  The figure shows the input/output
  relationships among the current set of tools in the Crux toolkit.
  \label{figure:schematic}}
\end{figure}

The first time a user runs Crux, they see a flow diagram showing the
input/output relationships among the various Crux tools, as in
Figure~\ref{figure:schematic}.  The user is asked to click on the
tools that they would like to use.  Clicking on a tool ``lights it
up'' in some fashion.  After lighting up one tool, clicking on a
mutually exclusive tool (i.e., clicking first percolator and then
compute-q-values) causes the first tool to go dark again.  Clicking on
a tool a second time also causes it to go dark.

When a tool is selected, input boxes for the required input files and
for several key parameters are displayed prominently, directly below
the schematic.  Optionally, the user can click on a ``Show additional
parameters'' button to display the complete list of parameters, with
their default values specified.

In addition to tool-specific parameters, Crux provides an option to
specify a name for this particular analysis.  Specifying such a name
will make it easier to retrieve the results of this analysis later.
If the user does not specify a name, then a name is created
automatically, using e.g., the name of the MS2 file and the date.

Once the user specifies, at the least, the inputs for the search
routine, the analysis can begin.  A ``Start analysis'' button appears
on the screen.  The user can continue modifying parameters as long as
they want, but they are free to start the analysis at any time.  At
that point, the analysis will begin, and the user is directed to an
output page.

The final output of the analysis is an XML ``summary file'' containing
links to the various tool outputs.  Initially, the summary file is
just a skeleton showing where the results will eventually appear, with
each tool listed as a link.  Links become live as soon as a tool
starts running, and the user can click on the link to see the current
progress of the job, or the final results.  Internally, each tool
creates its own directory, storing necessary parameters therein.  The
file simply links to these directories.  The file can be accessed via
a browser to view the results of the analysis, or the file can be
provided as input to Crux to initialize a subsequent analysis.

\subsection{Modifying a previous analysis}

To re-run an analysis, the user starts Crux and provides as input the
summary file produced by the previous analysis.  The Crux flow diagram
is displayed again, and boxes corresponding to analyses that have
already been run are indicated visually in some way.  Thus, in
general, each box takes one of three states: unselected, selected but
not run yet, or complete.

Most commonly, a user will want to re-run the same analysis on new
spectra.  In this case, they click on the ``search-for-matches'' box
and enter the name of a new MS2 file.  At this point, that box
switches to the ``not run yet'' state, and all of the downstream boxes
do as well.  Selected parameter settings remain the same.

Prior to re-running the analysis, the user must select a new name for
the experiment.  If they do not do so, then they will be asked whether
it is OK to delete the results of the previous analysis.  If the user
says ``yes,'' then the directories associated with the modified
analysis steps are deleted prior to starting the analysis.  If the
user says ``no,'' then the user is given the option of providing a new
name for the experiment.

\subsection{Robustness}

Crux assumes that the user will not disrupt the directory structure
that the toolkit creates.  The structure itself is quite simple: each
analysis consist of one summary file, plus a set of linked
directories, one per tool.  It is OK for a user to delete or move one
of these directories, but if they try to run Crux using the summary
file, the tool corresponding to the missing directory will be
indicated as ``not run yet,'' as will all downstream tools.
Similarly, if the user moves the summary file without moving the
corresponding tool-specific directories, then when Crux is run using
that summary file, all of the analyses will be displayed as ``not run
yet.''  Note that, since the parameters are all stored in the relevant
subdirectories, deleting or moving a tool-specific subdirectory will
result in re-setting all the parameters to their default values.

In general, the name that the user provides will be the name of the
summary file (with .xml appended).  Thus, re-running the same analysis
multiple times requires clobbering the corresponding output file and
output directories.  The GUI maintains internal logic indicating the
status of each job (not-run-yet, to-be-run, unselected).  Prior to any
analysis, the GUI checks whether any of the to-be-run jobs already
have output directories associated with them.  In this case, the user
is queried, as described above, whether to delete the problematic
directories or re-name the analysis.

Power users might want to cobble together different analyses, by
selecting output directories from different tools and creating their
own summary script.  They do so at their own risk.  In such a
scenario, the settings from one directory may not be compatible with
the settings in a downstream directory.

\subsection{A common single-tool interface}

The Crux summary page created by a single analysis contains links to
output directories created by each tool.  Clicking on any of these
links leads to a tool-specific summary page.  All tools create summary
pages that share a common ``look and feel.''  Each such page contains
a brief summary of the analysis, with links to a parameter file,
a log file, the input files and each of the output files.

The indexing tool does not create user-level output files.  Instead,
the summary page simply lists some basic statistics about the index:
the total number of proteins in the index, the average protein length,
the total number of peptides in the index, the average number of
peptides per protein, the average peptide length.

\subsection{Manipulating ranked lists}

Most Crux tools create, as their primary output, a ranked list of
objects---PSMs, peptides or proteins.  Eventually, we would like to
create an interface that allows users to manipulate these ranked lists
in the following ways:
\begin{itemize}
\item Features (columns) can be hidden or displayed, at the user's
  discretion.
\item The list can be sorted according to the values in one or more
  columns, in either ascending or descending order.
\item The list can be thresholded according to the values in one or
  more columns, using either a maximum or minimum value.
\item A protein-level ranking can be expanded (either globally or on a
  per-protein basis) to list peptides, and peptides can in turn be
  expanded to list individual spectra.
\item Spectra can optionally be expanded to display clickable
  thumbnail images, with various types of peaks (b- and y-ions)
  annotated in color.
\end{itemize}
Because creating such an interface promises to be onerous, this
activity will be shelved for the foreseeable future.

\section{Implementation}

\subsection{Introduction}

The Crux GUI will be implemented

\begin{itemize}
\item as a desktop app 
\item based on the Model-View-Controller (MVC) pattern
\item using Java SE 6 and the SWING framework
\end{itemize}

Implementation as a desktop application provides the user with the 
most independence, and avoids the need for a dedicated Crux web server.
MVC is the most common architecture for GUI applications.
In the MVC pattern 
\begin{itemize}
\item The Model is the repository for the underlying data.
\item The View is responsible for constructing the visual representation of the Model,
and manages interaction with the user.
\item The Controller acts as an intermediary between a View and a Model:
the View notifies the Controller of user requests to modify the Model,
the Controller carries out the modification of the model,
and notifies the View(s) of the completed changes to the Model.
\end{itemize}
While the names are singular, multiple Models, Views, and Controllers
may be in operation simultaneously.

There are circumstances where the View doesn't need to modify the Model.
In these cases the Controller and the View can be merged into a single component.
This is referred to as a Model-Delegate pattern. The component combining 
the controller and the View is called a UI-delegate.

Java SE 6 was selected because it is ubiquitous on Linux, Mac OS X, and Windows.
We investigated the possibility of implementing the Crux GUI using HTML-CSS 
markup and Javascript,
but the Javascript security model does not allow local libraries or executables to
be invoked directly from Javascript.
Java applications allow a very simple installation process.
Installation may be as simple as expanding an archive containing 
jar files and shell scripts and batch files for launching the JVM.
Java Web start provides a particularly simple method of installing Java applications
and the Java Runtime Environment (JRE) but because the Crux GUI needs access to the local file
system we would need to sign the application, and supporting a signed application
may require some infrastructure in setting up the certificates.
If necessary, we can redistribute the JRE as part of the Crux package without affecting
the Crux license.

\subsection{Model}

The following classes will encapsulate data:

\begin{itemize}
\item Analysis -- all parameters needed to perform an analysis
\item Summary -- summary of analysis results
\item Protein
\item Peptide
\item Spectrum
\item PeptideSpectrumMatch
\end{itemize}

The Model classes can be aggregated using the Java container classes or subclasses
inheriting from the container classes. The Java String class is the Model for
plain text.

The Analysis class will track the parameters in the Table~1.
\begin{table}[b]
\scriptsize
\begin{tabular}{lllll}
\hline
\multicolumn{5}{c}{Basic Parameters} \\
\hline
\textbf{Index}&\textbf{Search}&\textbf{Compute q-values}&\textbf{Percolator} &\textbf{Q-Ranker} \\
\hline \\
digestion            &                                  &                                 &                                 &                                 \\
enzyme               &decoy-p-values                    &protein input                    &protein input                    &protein input                    \\
isotopic-mass        &                                  &                                 &                                 &                                 \\
missed-cleavages     &                                  &                                 &                                 &                                 \\
protein fasta file   &ms2 file                          &                                 &                                 &                                 \\
                     &num-decoy-files                   &                                 &                                 &                                 \\
                     &num-decoys-per-target             &                                 &                                 &                                 \\
                     &protein input                     &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
\hline
\multicolumn{5}{c}{Advanced Parameters} \\
\hline \\
custom-enzyme        &print-search-progress             &pi0                              &feature-file                     &feature-file                     \\
max-length           &                                  &                                 &                                 &pi0                              \\
max-mass             &                                  &qvalues-tab-output-file          &                                 &                                 \\
min-length           &                                  &                                 &pi0                              &                                 \\
min-mass             &seed                              &                                 &top-match                        &top-match                        \\
                     &spectrum-charge                   &                                 &                                 &                                 \\
                     &spectrum-max-mass                 &                                 &                                 &                                 \\
                     &spectrum-min-mass                 &                                 &                                 &                                 \\
                     &max-mods                          &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
                     &                                  &                                 &                                 &                                 \\
\end{tabular}
\caption{Basic and advanced parameters}
\end{table}

compute-p-values will always be set to True
parameter-file will always be set to True
\subsection{View}

The following classes will implement Views:

\begin{itemize}
\item ParentView -- this View is the parent in the visual hierarchy for all the other viewers. 
It contains the "always on" controls (minimize, maximize, help, exit for example).
\item AnalysisView -- this Viewer presents the Crux analysis diagram and
allows the user to specify the parameters of the analysis.
Only the basic parameters will be shown be default. 
A button will be provided to turn on the display of the advanced parameters listed in
Table~1.
\item SummaryView -- this Viewer presents any of the Summary Models
and allows the user to select linked items for display.
\end{itemize}

\subsection{Controller}

The following classes will implement Controllers:

\begin{itemize}
\item CruxMasterController -- contains the entry
point for the program and manages the other Controllers, Viewers, and UI-Delegates.
It executes Crux when requested and performs the actions requested from any of 
the "always on" controls.
\item AnalysisViewController -- this Controller mediates between
the Analysis and the AnalysisView. It validates and executes the 
requested actions from the Analysis viewer, modifying the Analysis Model
as necessary.
\item SummaryViewController -- this Controller mediates between
Summary and SummaryView.
\end{itemize}

\subsection{UI Delegates}

The following classes will implement UI Delegates. They correspond to Views
that don't need to modify the underlying Model.

\begin{itemize}
\item TextView 
\item RankedListView
\item SpectrumView
\end{itemize}

\end{document}
